architecture.md

```diff
@@ -0,0 +1,56 @@
+# Architecture of Metals
+This document describes the high-level architecture of Metals following the philosophy of [rust-analyzer's ARCHITECTURE.md](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html)
+
+## Other resources
+- For the general contributing guideline, see [Getting started guide for contributors](https://scalameta.org/metals/docs/contributors/getting-started/).
+- If you're interested in the overview of how Metals works, [A Dive into how Metals works](https://youtu.be/fpzN_vTBy18) video will be helpful.
+- For LSP specification, please refer [the LSP document](https://microsoft.github.io/language-server-protocol/), you don't have to read all of the `specification`, just read the `overview` and pick out the interesting parts when you need it.
+
+## LSP endpoints
+
+The most important file of this project is `MetalsLanguageServer.scala`. In this file, we define all of the LSP endpoints, for example, you will find the endpoint for `textDocument/completion` request as follows.
+
+```scala
+@JsonRequest("textDocument/completion")
+def completion(...) = ...
+```
+
+In addition to the LSP endpoints, this file contains many components, for example `private var semanticdbs: Semanticdbs = _`. Those components will be initialized later at some points like on LSP's `initialize` request.
+
+## Presentation compiler
+Metals features are powered by presentation compilers, if you hit `Compilers.scala` it is the client of compilers.
+
+`ScalaPresentationCompiler.scala` under `scala-2` and `scala-3` works as a wrapper of Scala compilers.
+
+For more details
+
+- [An introduction to the Scala presentation compiler](https://www.chris-kipp.io/blog/an-intro-to-the-scala-presentation-compiler)
+- Presentation compilers' endpoint
```

Comment:
There are endpoint from the compiler, so maybe explain that metals contacts those endpoint via those classes?